components/crypto_dot_com/browser/crypto_dot_com_service.cc

```diff
@@ -0,0 +1,242 @@
+    GetAssetRankingsCallback callback) {
+  auto internal_callback = base::BindOnce(
+      &CryptoDotComService::OnAssetRankings,
+      base::Unretained(this), std::move(callback));
+  GURL url = GetURLWithPath(root_host, get_gainers_losers_path);
+  return NetworkRequest(
+      url, "GET", "", std::move(internal_callback));
+}
+
+void CryptoDotComService::OnAssetRankings(
+    GetAssetRankingsCallback callback,
+    const int status, const std::string& body,
+    const std::map<std::string, std::string>& headers) {
+  CryptoDotComAssetRankings rankings;
+  if (status >= 200 && status <= 299) {
+    const std::string json_body = GetFormattedResponseBody(body);
+    CryptoDotComJSONParser::GetRankingsFromJSON(json_body, &rankings);
+  }
+  std::move(callback).Run(rankings);
+}
+
+bool CryptoDotComService::NetworkRequest(const GURL &url,
+                                  const std::string& method,
+                                  const std::string& post_data,
+                                  URLRequestCallback callback) {
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = url;
+  request->load_flags = net::LOAD_BYPASS_CACHE |
+                        net::LOAD_DISABLE_CACHE |
+                        net::LOAD_DO_NOT_SEND_COOKIES |
+                        net::LOAD_DO_NOT_SAVE_COOKIES;
+  request->method = method;
+
+  auto url_loader = network::SimpleURLLoader::Create(
+      std::move(request), GetNetworkTrafficAnnotationTag());
+
+  if (!post_data.empty()) {
+    url_loader->AttachStringForUpload(post_data,
+        "application/x-www-form-urlencoded");
+  }
+
+  url_loader->SetRetryOptions(
+      kRetriesCountOnNetworkChange,
+      network::SimpleURLLoader::RetryMode::RETRY_ON_NETWORK_CHANGE);
+
+  auto iter = url_loaders_.insert(url_loaders_.begin(), std::move(url_loader));
+  auto* default_storage_partition =
+      content::BrowserContext::GetDefaultStoragePartition(context_);
+  auto* url_loader_factory =
+      default_storage_partition->GetURLLoaderFactoryForBrowserProcess().get();
+
+  iter->get()->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
+      url_loader_factory, base::BindOnce(
+          &CryptoDotComService::OnURLLoaderComplete,
+          base::Unretained(this), std::move(iter), std::move(callback)));
+
+  return true;
+}
+
+void CryptoDotComService::OnURLLoaderComplete(
+    SimpleURLLoaderList::iterator iter,
+    URLRequestCallback callback,
+    const std::unique_ptr<std::string> response_body) {
+  auto* loader = iter->get();
+  auto response_code = -1;
+  std::map<std::string, std::string> headers;
+  if (loader->ResponseInfo() && loader->ResponseInfo()->headers) {
+    response_code = loader->ResponseInfo()->headers->response_code();
+    auto headers_list = loader->ResponseInfo()->headers;
+    if (headers_list) {
+      size_t iter = 0;
+      std::string key;
+      std::string value;
+      while (headers_list->EnumerateHeaderLines(&iter, &key, &value)) {
```

Comment:
nit: Future cleanup not needed to be part of this PR, I think this block of code is repeated in a few places.

Reply:
We can put this in a shared utility for sure